docs: add "Building a middleware from scratch" guide (#590)
tower-rs/tower: guides/building-a-middleware-from-scratch.md

# Building a middleware from scratch

In ["Inventing the `Service` trait"][invent] we learned all the motivations
behind [`Service`] and why its designed the way it is. We also built a few
smaller middleware ourselves but we took a few shortcuts in our implementation.
In this guide we're going to build the `Timeout` middleware as it exists in
Tower today without taking any shortcuts.

Writing a robust middleware requires working with async Rust at a slightly lower
level than you might be used to. The goal of this guide is to demystify the
concepts and patterns so you can start writing your own middleware and maybe
even contribute back to the Tower ecosystem!

## Getting started

The middleware we're going to build is [`tower::timeout::Timeout`]. It will set

> a limit on the maximum duration its inner `Service` is allowed to take. If it

Reviewer: I would maybe say it in this way, the maximum duration its inner service can dispatch for? Since its not the duration of the service but the response returned from `call`?

doesn't produce a response within some amount of time, an error is returned.
This allows the client to retry that request or report an error to the user,
rather than waiting forever.

Lets start by writing a `Timeout` struct that holds the `Service` its wrapping
and the duration of the timeout:

```rust
use std::time::Duration;

struct Timeout<T> {
    inner: T,
    timeout: Duration,
}
```

> As we learned in ["Inventing the `Service` trait"][invent] its important for

Reviewer: Hmm Im not sure if this is important it is just one method. The key point is that there are sometimes things you need to own for the response, ie the duration of the timeout but I believe even in timeout you can create the timeout future then move that in. So why would the service need to be clone (besides that if it can impl clone it should to forward). So I guess my point is that I think this circumvents why clone is important for service middleware rather than using it for moving the data it should be there to forward (and sometimes you do need to move self but not always).

Author: Yes, `Timeout` doesn't require its service to be Clone but I still think the general recommendation of making services Clone, so services can be moved into response futures, is important. So I think I'm going to leave this as is.

services to implement `Clone` such that you can convert the `&mut self` given to
`Service::call` into an owned `self` that can be moved into the response future,
if necessary. We should therefore add `#[derive(Clone)]` to our struct. We

> should also drive `Debug` while we're at it:

Reviewer 2: ```suggestion
should also derive `Debug` while we're at it:
```

```rust

> #[derive(Debug, Clone)]

Reviewer 2: I think this is common enough in the ecosystem that I probably wouldn't even include text on why we're choosing to do it. See also [the Rust API guidelines' recommendation](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits).

Author: Hm yeah thats true. I think I'm going to leave it though since its extra important for services so users can move the service into the response future.

struct Timeout<S> {
    inner: S,
    timeout: Duration,
}
```

Next we write a constructor:

```rust
impl<S> Timeout<S> {
    pub fn new(inner: S, timeout: Duration) -> Self {
        Timeout { inner, timeout }
    }
}
```

> Note that we don't have to add any trait bounds on `S` even though we expect `S`

Reviewer: Good call out! 👍

Reviewer 2: I actually wonder if we should make this even briefer:
> Note that we omit bounds on `S` even though we expect it to implement `Service`, as the [Rust API guidelines recommend](https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-bounds).

Author: Yeah I like that 👍

    }
}
```

Until you've written lots of middleware writing out a skeleton like this makes
the process a bit easier.

To actually add a timeout to the inner service what we essentially have to do is
detect when the future returned by `self.inner.call(request)` has been running
longer than `self.duration` and abort with an error.

The approach we're going to take is to call [`tokio::time::sleep`] to get a
future that completes when we're out of time and then select the value from
whichever of the two futures is the first to complete. We could also use
`tokio::time::timeout` but `sleep` works just as well.

Creating both futures is done like this:

```rust
use tokio::time::sleep;

fn call(&mut self, request: Request) -> Self::Future {
    let response_future = self.inner.call(request);

    // This variable has type `tokio::time::Sleep`.
    //
    // We don't have to clone `self.duration` as it implements the `Copy` trait.
    let sleep = tokio::time::sleep(self.timeout);

    // what to write here?
}
```

One possible return type is `Pin<Box<dyn Future<...>>>`. However we want our
`Timeout` to add as little overhead as possible, so we would like to find a way

> to avoid allocating a `Box`. Imagine we have a large stack, with dozens of

Reviewer: Would be good to call out that `type_alias_impl_trait` fixes this issue since we can return an unnamed type.

Author: I've added a footnote about it. Generally not a fan of mentioning features that have been "coming soon" for years 😜

nested `Service`s, where each layer allocates a new `Box` for every request
that passes through it. That would result in a lot of allocations which might
impact performance.

## The response future

To avoid using `Box` lets instead write our own `Future` implementation. We
start by creating a struct called `ResponseFuture`. It has to be generic over
the inner service's response future type. This is analogous to wrapping services
in other services, but this time we're wrapping futures in other futures.

```rust
use tokio::time::Sleep;

> pub struct ResponseFuture<F> {

Reviewer 2: ```suggestion
pub struct FutureWithTimeout<F> {
```

Author: I actually think I'm going to stick with `ResponseFuture` as that is what its called in Tower. I quite like that people can read this guide and when they're done be familiar with the code in Tower.

Reviewer 2: Yeah, I suppose it's okay in Tower it's okay as the type lives in its own module.

```
error[E0599]: no method named `poll` found for type parameter `F` in the current scope
  --> src/lib.rs:56:29
   |
56 |         match self.response_future.poll(cx) {
   |                             ^^^^ method not found in `F`
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following traits define an item `poll`, perhaps you need to restrict type parameter `F` with one of them:
   |
49 | impl<F: Future, Response, Error> Future for ResponseFuture<F>
   |      ^^^^^^^^^

error: aborting due to previous error
```

Unfortunately the error we get from Rust isn't very good. It tells us to add a
`F: Future` bound even though we've already done that with `where F:
Future<Output = Result<Response, E>>`.

The real issue has to do with [`Pin`]. The full details of pinning is outside
the scope of this guide. If you're brand new to `Pin` we recommend ["The Why,
What, and How of Pinning in Rust"][pin] by Jon Gjengset.

What Rust is trying to tell us is that we need a `Pin<&mut F>` to be able to
call `poll`. Accessing `F` through `self.response_future` when `self` is a
`Pin<&mut Self>` doesn't work.

> One solution is to make `ResponseFuture<F>` implement [`Unpin`]. If a type

Reviewer 2: This feels like a level of detail that isn't necessary for this article. I would jump straight to `pin_project` without first going through this strawman.

Author: Hm yeah thats probably true. I'm going to edit things down.

implements `Unpin` it means we can ignore `Pin` and treat a `Pin<&mut Self>` as
just a `&mut self`, as if the `Pin` wasn't there. As `Unpin` is an auto trait
the compiler will automatically implement it if all types contained are also

> `Unpin`, similarly to `Send` and `Sync`.

Reviewer: another interesting call out is that implementing `Unpin` manually is safe while doing `Send`/`Sync` is `unsafe impl`. This is because `Unpin` is safe to `impl` on a struct as long as you don't use any other unsafe code to access stuff. Its not unsafe but that is where the unsaftey comes up. So if you can impl `Unpin` on your struct and not use any `unsafe` to get any of the pins out then its totally safe. Its a bit different from Send/Sync in this case.

Author: Yep! That is something Jon covers in his video so don't think I'll mention it here.

So to make `ResponseFuture<F>` implement `Unpin` we need `F` and `Sleep` to
implement `Unpin`. However `tokio::time::Sleep` is not `Unpin`. So even if we
made `F` somehow implement `Unpin`, `ResponseFuture<F>` still wouldn't be
`Unpin`.

Its possible to make any type `Unpin` using `Box::pin`. But that requires
allocating which is what we're trying to avoid. The proper solution is to use
"pin projection".

> Pin projection sounds a lot more fancy than it really is. It basically means to

Reviewer 3: ```suggestion
Pin projection sounds a lot fancier than it really is. It basically means to
```

finishes first?

## The error type

The error type we're promising to return now is the generic `Error` type which
is the same as the inner service's error type. However we know nothing about
that type. It is completely opaque to us and we have no way of constructing
values of that type.

We have three options:

1. Return a boxed error trait object like `Box<dyn std::error::Error + Send +
   Sync>`.
2. Return an enum with variants for the service error and the timeout error.
3. Define a `TimeoutError` struct and require that our generic error type can be
   constructed from a `TimeoutError` using `TimeoutError: Into<Error>`.

While option 3 might seem like it is the most flexible it isn't great since it
requires users using a custom error type to manually implement
`From<TimeoutError> for MyError`. That quickly becomes tedious when using lots
of middleware that each have their own error type.

Option 2 would mean defining an enum like this:

```rust
enum TimeoutError<Error> {
    // Variant used if we hit the timeout
    Timeout(InnerTimeoutError),
    // Variant used if the inner service produced an error
    Service(Error),
}
```

> While this seems ideal on the surface as we're not loosing any type information

Reviewer 3: ```suggestion
While this seems ideal on the surface as we're not losing any type information
```

advantages:

1. Our error handling is less fragile since changing the order middleware are
   applied in wont change the final error type.
2. The error type now has a constant size regardless how many middleware we've
   applied.
3. Extracting the error no longer requires a big `match` but can instead be done
   with `error.downcast_ref::<Timeout>()`.

However it also has the following downsides:

1. As we're using dynamic downcasting the compiler can no longer guarantee that
   we're exhaustively checking for every possible error type.
2. Creating an error now requires an allocation. In practice we expect errors to
   be infrequent and therefore this shouldn't be a problem.

Which option you prefer is a matter of personal preference. Both have their
advantages and disadvantages. However the pattern that we've decided to use in
Tower is boxed trait objects. You can find the original discussion
[here](https://github.com/tower-rs/tower/issues/131).

For our `Timeout` middleware that means we need to create a struct that
implements `std::error::Error` such that we can convert it into a `Box<dyn
std::error::Error + Send + Sync>`. We also have to require that the inner

> service's error type implements `Into<Box<dyn std::error::Error + Send +

Reviewer: And one note here is that we must always use `Into` if some middleware use `From` and other use `Into` it will break the stack in horrible to understand ways.

Author: Good point! std mentions that. Gonna add a link to it.

Sync>>`. Luckily most errors automatically satisfies that so it wont require
users to write any additional code.

The code for our error type looks like this:

```rust
use std::fmt;

#[derive(Debug, Default)]
pub struct TimeoutError(());

impl fmt::Display for TimeoutError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.pad("request timed out")
    }
}

impl std::error::Error for TimeoutError {}
```

We add a private field to `TimeoutError` such that users outside of Tower cannot
construct their own `TimeoutError`. They can only be obtained through our

> middleware.

Reviewer 2: I think the "modern" way of doing this is with `#[non_exhaustive]`.

Author: That doesn't seem to work for empty structs https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ccc7ef1027273fdb59abe545aa20512

Reviewer 2: `#[non_exhaustive]` is only enforced across crate boundaries :)

Author: Ah yes of course! Cool. Didn't know 😊

`Box<dyn std::error::Error + Send + Sync>` is also quite a mouthful so
lets define a type alias for it:

```rust
// This also exists as `tower::BoxError`
pub type BoxError = Box<dyn std::error::Error + Send + Sync>;
```

Our future implementation now becomes:

```rust
impl<F, Response, Error> Future for ResponseFuture<F>
where
    F: Future<Output = Result<Response, Error>>,
    // Require that the inner service's error can be converted into a `BoxError`.
    Error: Into<BoxError>,
{
    type Output = Result<
        Response,
        // The error type of `ResponseFuture` is now `BoxError`.
        BoxError,
    >;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let this = self.project();

        match this.response_future.poll(cx) {
            Poll::Ready(result) => {
                // Use `map_err` to convert the error type.

>                 let result = result.map_err(Into::into);

Reviewer 2: I _think_ you can just use `?` directly on `Poll<Result<_, _>>` now?

Author: Yeah I think you're right. I'm going to leave it as is because I think the additional verbosity might help those who haven't written futures before.
